File: problems/026/026.f90
PROGRAM F026

    ! Copyright 2021 Melwyn Francis Carlo

    IMPLICIT NONE

    CHARACTER (LEN=2000) :: BASE_NUMBER
    CHARACTER (LEN=8)    :: BASE_SUB_NUMBER

    CHARACTER (LEN=:), ALLOCATABLE :: TEMP_SUB_STR_001, TEMP_SUB_STR_002
    CHARACTER (LEN=:), ALLOCATABLE :: TEMP_SUB_STR_01,  TEMP_SUB_STR_02

    INTEGER :: I_MAX = 0
    INTEGER :: N     = 80

    INTEGER :: I, J, CARRY_VAL, FIND_LENGTH, QUOTIENT

    TEMP_SUB_STR_01  = ""
    TEMP_SUB_STR_02  = ""
    TEMP_SUB_STR_001 = ""
    TEMP_SUB_STR_002 = ""

    DO I = 11, 999

        IF (MOD(I, 2) == 0) CYCLE
        IF (MOD(I, 5) == 0) CYCLE
        IF (MOD(I, 7) == 0) CYCLE

        BASE_NUMBER = REPEAT("0", 2000);
        BASE_NUMBER(1:1) = '1';

        CARRY_VAL = 0
        DO J = 1, 2000, 5

            BASE_SUB_NUMBER = ACHAR(0)
            WRITE (BASE_SUB_NUMBER, '(I0.3)') CARRY_VAL
            BASE_SUB_NUMBER(4:8) = BASE_NUMBER(J:J+4)
            READ (BASE_SUB_NUMBER, *) QUOTIENT
            CARRY_VAL = MOD(QUOTIENT, I)
            QUOTIENT  = FLOOR (FLOAT(QUOTIENT) / FLOAT(I))

            WRITE (BASE_SUB_NUMBER, '(I0.5)') QUOTIENT
            BASE_NUMBER(J:J+4) = BASE_SUB_NUMBER(1:5)

        END DO

        J = 1
        DO WHILE (BASE_NUMBER(J:J) == '0')
            J = J + 1
        END DO

        FIND_LENGTH = N;

        DO WHILE (J <= (2001 - FIND_LENGTH))

            DEALLOCATE(TEMP_SUB_STR_001)
            ALLOCATE(CHARACTER (LEN=FIND_LENGTH) :: TEMP_SUB_STR_001)
            TEMP_SUB_STR_001 = BASE_NUMBER(J:J+FIND_LENGTH-1)

            DEALLOCATE(TEMP_SUB_STR_002)
            ALLOCATE(CHARACTER (LEN=(2001-FIND_LENGTH-J)) :: TEMP_SUB_STR_002)
            TEMP_SUB_STR_002 = BASE_NUMBER(J+FIND_LENGTH:2000)

            IF (INDEX(TEMP_SUB_STR_002, TEMP_SUB_STR_001) /= 0) THEN

                DEALLOCATE(TEMP_SUB_STR_01)
                ALLOCATE(CHARACTER (LEN=10) :: TEMP_SUB_STR_01)
                TEMP_SUB_STR_01 = TEMP_SUB_STR_001(1:10)

                DEALLOCATE(TEMP_SUB_STR_02)
                ALLOCATE(CHARACTER (LEN=(FIND_LENGTH-10)) :: TEMP_SUB_STR_02)
                TEMP_SUB_STR_02 = TEMP_SUB_STR_001(11:FIND_LENGTH)

                IF (INDEX(TEMP_SUB_STR_02, TEMP_SUB_STR_01) == 0) THEN

                    N = FIND_LENGTH
                    FIND_LENGTH = FIND_LENGTH + 1

                    IF (TEMP_SUB_STR_002(INDEX(TEMP_SUB_STR_002, TEMP_SUB_STR_001):) == TEMP_SUB_STR_002) THEN
                        I_MAX = I
                        EXIT
                    ELSE
                        CYCLE
                    END IF

                END IF

            END IF

            J = J + 1

        END DO

    END DO

    PRINT ('(I0)'), I_MAX

END PROGRAM F026

File: problems/026/026.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Copyright 2021 Melwyn Francis Carlo */

/* A simpler solution is to find the greatest prime number less than 1000  */
/* that is a safe prime and a full reptend.                                */

int main()
{
    int i_max = 0;
    int n     = 80;

    for (int i = 11; i < 1000; i++)
    {
        if ((i % 2) == 0)
            continue;

        if ((i % 5) == 0)
            continue;

        if ((i % 7) == 0)
            continue;

        char base_number[2001];
        memset(&base_number[0], '0', 2000);
        base_number[2000] = 0;
        base_number[0] = '1';

        int carry_val = 0;
        for (int j = 0; j < 2000; j += 10)
        {
            char base_sub_number[14] = { 0 };
            sprintf(&base_sub_number[0], "%03d", carry_val);
            memcpy(&base_sub_number[3], &base_number[j], 10);
            long int quotient = (long int)(atol(base_sub_number) / i);
            carry_val = atol(base_sub_number) % i;

            sprintf(&base_sub_number[0], "%010lu", quotient);
            memcpy(&base_number[j], &base_sub_number[0], 10);

        }

        int j = 0;
        while (base_number[j] == '0')
            j++;

        int find_length = n;

        while (j <= (2000 - find_length))
        {
            char temp_sub_str_001[find_length+1];
            temp_sub_str_001[find_length] = 0;
            memcpy(&temp_sub_str_001[0], &base_number[j], find_length);

            char temp_sub_str_002[2001-find_length-j];
            temp_sub_str_002[2000-find_length-j] = 0;
            memcpy(&temp_sub_str_002[0], &base_number[j+find_length], 2000 - find_length - j);

            if (strstr(temp_sub_str_002, temp_sub_str_001) != NULL)
            {
                const int DIGITS_LEN = 10;

                char temp_sub_str_01[DIGITS_LEN+1];
                temp_sub_str_01[DIGITS_LEN] = 0;
                memcpy(&temp_sub_str_01[0], &temp_sub_str_001[0], DIGITS_LEN);

                char temp_sub_str_02[find_length-DIGITS_LEN+1];
                temp_sub_str_02[find_length-DIGITS_LEN] = 0;
                memcpy(&temp_sub_str_02[0], &temp_sub_str_001[DIGITS_LEN], find_length - DIGITS_LEN);

                if (strstr(temp_sub_str_02,  temp_sub_str_01) == NULL)
                {
                    n = find_length++;

                    if ((strstr(temp_sub_str_002, temp_sub_str_001) - temp_sub_str_002) == 0)
                    {
                        i_max = i;
                        break;
                    }
                    else
                    {
                        continue;
                    }
                }
            }

            j++;
        }
    }

    printf("%d\n", i_max);

    return 0;
}
